### PR TITLE
types: use api-types for stricter typings

### PR DESCRIPTION
**Pull request by ckohen**
**Please describe the changes this PR makes and why it should be merged:**

-types is now a real dep so lets use it to the fullest

(sorry about the diff...its kinda massive, tried to make it easier by separating the raw tpes)

**Status and versioning classification:**

- I know how to update typings and have done so, or typings don't need updating

### CODE REVIEW

**Comment by ckohen** [comment]
And the instant rebase has been done

File: typings/index.d.ts [hunk 1]
```diff
@@ -238,11 +303,12 @@ export class BitField<S extends string, N extends number | bigint = number> {
   public toJSON(): N extends number ? number : string;
   public valueOf(): N;
   public [Symbol.iterator](): IterableIterator<S>;
-  public static FLAGS: unknown;
+  public static FLAGS: Record<string, number | BigInt>;
```
Comment (kyranet) [inline typings/index.d.ts:306]:
```suggestion
  public static FLAGS: Record<string, number | bigint>;
```

File: typings/index.d.ts [hunk 2]
```diff
@@ -973,9 +1045,15 @@ export class LimitedCollection<K, V> extends Collection<K, V> {
...
-  private _update(data: unknown, partial?: boolean): Message;
+  public constructor(
+    client: Client,
+    data: RawMessageData,
+    channel: TextChannel | DMChannel | NewsChannel | ThreadChannel,
+  );
+  private _patch(data: RawPartialMessageData, partial: true): Message;
+  private _patch(data: RawMessageData): Message;
+  private _update(data: RawPartialMessageData, partial: true): Message;
+  private _update(data: RawMessageData): Message;
```
Comment (kyranet) [inline typings/index.d.ts:1056]:
```suggestion
  private _patch(data: RawPartialMessageData, partial: true): Message;
  private _patch(data: RawMessageData, partial?: false): Message;
  private _update(data: RawPartialMessageData, partial: true): Message;
  private _update(data: RawMessageData, partial?: false): Message;
```

File: typings/index.d.ts [hunk 3]
```diff
@@ -2573,7 +2662,7 @@ export interface WebhookFields extends PartialWebhookFields {
   readonly createdTimestamp: number;
   delete(reason?: string): Promise<void>;
   edit(options: WebhookEditData, reason?: string): Promise<Webhook>;
-  sendSlackMessage(body: object): Promise<boolean>;
+  sendSlackMessage(body: unknown): Promise<boolean>;
```
Comment (kyranet) [inline typings/index.d.ts:2665]: Is it not possible to type this?

**Review by kyranet — CHANGES_REQUESTED** [review]
I love this, just a few changes:

File: typings/index.d.ts (hunk 3 above)
Comment (vladfrangu) [reply typings/index.d.ts:2665]: We'd need to type it off of Slack's documentation..

File: typings/index.d.ts [hunk 4]
```diff
@@ -2800,9 +2889,9 @@ export type ApplicationFlagsString =
   | 'EMBEDDED';
 
 export interface AuditLogChange {
-  key: string;
-  old?: unknown;
-  new?: unknown;
+  key: APIAuditLogChange['key'];
+  old?: APIAuditLogChange['old_value'];
+  new?: APIAuditLogChange['new_value'];
 }
```
Comment (vladfrangu) [inline typings/index.d.ts:2895]: Does this... pick up the unions? If it does, that's awesome!

File: typings/index.d.ts [hunk 5]
```diff
@@ -1292,19 +1371,20 @@ export class NewsChannel extends TextBasedChannel(GuildChannel) {
...
 
 export class OAuth2Guild extends BaseGuild {
+  constructor(client: Client, data: RawOAuth2GuildData);
```
Comment (vladfrangu) [inline typings/index.d.ts:1374]:
```suggestion
  public constructor(client: Client, data: RawOAuth2GuildData);
```

File: typings/index.d.ts (hunk 1 above)
Comment (vladfrangu) [reply typings/index.d.ts:306]: Shouldn't this be `Record<S, bigint>` instead?

**Review by vladfrangu — DISMISSED** [review]

File: typings/index.d.ts (hunk 1 above)
Comment (kyranet) [reply typings/index.d.ts:306]: This is BitField, which can be either.

Comment (vladfrangu) [reply typings/index.d.ts:306]: Disregarding that bigint part, the key is still relevant

Comment (NotSugden) [reply typings/index.d.ts:306]: `FLAGS` is static so you can't use the type parameter 😭

File: typings/index.d.ts (hunk 3 above)
Comment (iCrawl) [reply typings/index.d.ts:2665]: Yeah let's not.

File: typings/index.d.ts (hunk 4 above)
Comment (ckohen) [reply typings/index.d.ts:2895]: It does! It's quite awesome
image
image

**Comment by advaith1** [comment]
why are some types in rawDataTypes just re-exports from d-api-types? cant those be used directly?

**Comment by ckohen** [comment]
Because I wanted one place for all raw data types, in the event that any of them change it changes in one place and not multiple. Technically speaking I de-duplicated a lot of that, most of the ones that simply seem like reexports are deduplicated versions of a gateway event and 5 different rest endpoint results.

**Review by SpaceEEC — APPROVED** [review]

**Review by kyranet — APPROVED** [review]

**Review by iCrawl — APPROVED** [review]

**Review by vladfrangu — APPROVED** [review]

**Pull request merged**